File: app/src/main/java/be/appfoundry/aipdemo/activity/rxjava/ReactiveActivity.java
package be.appfoundry.aipdemo.activity.rxjava;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.widget.Button;
import android.widget.TextView;

import com.jakewharton.rxbinding2.view.RxView;

import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import be.appfoundry.aipdemo.AIPDemoApplication;
import be.appfoundry.aipdemo.R;
import be.appfoundry.aipdemo.model.StarWarsFilm;
import be.appfoundry.aipdemo.model.StarWarsFilms;
import be.appfoundry.aipdemo.service.SwapiService;
import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.functions.Predicate;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;

public class ReactiveActivity extends AppCompatActivity {

    @BindView(R.id.activity_reactive_log) TextView logView;
    @BindView(R.id.activity_reactive_button) Button button;

    @Inject SwapiService swapiService;

    private String log = "";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_reactive);

        ButterKnife.bind(this);

        AIPDemoApplication.getAppComponent().inject(this);

        simpleRxJava();
        //simpleRxJavaWithOperators();
        //combinedObservables();
        //reactiveButton();
        //reactiveDebounceButton();
        //reactiveService();
        //reactiveServiceOneByOne();
    }

    private void simpleRxJava() {
        Observable<String> simpleStringObservable = Observable.just("Eric", "Koen", "Bart", "Lies", "Niki");

        Observer<String> simpleStringObserver = new DisposableObserver<String>() {
            @Override
            public void onNext(String s) {
                logOnNext(s);
            }

            @Override
            public void onError(Throwable e) {
                logOnError(e);
            }

            @Override
            public void onComplete() {
                logOnCompleted();
            }
        };

        simpleStringObservable.subscribe(simpleStringObserver);
    }

    private void simpleRxJavaWithOperators() {
        Observable<String> simpleStringObservable = Observable.just("Eric", "Koen", "Bart", "Lies", "Niki");

        Observer<String> simpleStringObserver = new DisposableObserver<String>() {
            @Override
            public void onNext(String s) {
                logOnNext(s);
            }

            @Override
            public void onError(Throwable e) {
                logOnError(e);
            }

            @Override
            public void onComplete() {
                logOnCompleted();
            }
        };

        simpleStringObservable
                .filter(new Predicate<String>() {
                    @Override
                    public boolean test(String s) throws Exception {
                        return s.contains("i");
                    }
                })
                .map(new Function<String, String>() {
                    @Override
                    public String apply(String s) throws Exception {
                        return s.toUpperCase();
                    }
                })
                .subscribe(simpleStringObserver);
    }

    private void combinedObservables() {
        Observable<String> simpleMenObservable = Observable.just("Eric", "Koen", "Bart");
        Observable<String> simpleWomenObservable = Observable.just("Lies", "Niki");

        Observer<String> simpleStringObserver = new DisposableObserver<String>() {
            @Override
            public void onNext(String s) {
                logOnNext(s);
            }

            @Override
            public void onError(Throwable e) {
                logOnError(e);
            }

            @Override
            public void onComplete() {
                logOnCompleted();
            }
        };

        Observable.concat(simpleMenObservable, simpleWomenObservable).subscribe(simpleStringObserver);
    }

    private void reactiveButton() {
        Observable<Object> clickObservable = RxView.clicks(button);

        Observer<Object> clickObserver = new DisposableObserver<Object>() {
            @Override
            public void onNext(Object o) {
                logOnNext("Click");
            }

            @Override
            public void onError(Throwable e) {
                logOnError(e);
            }

            @Override
            public void onComplete() {
                logOnCompleted();
            }
        };

        clickObservable.subscribe(clickObserver);
    }

    private void reactiveDebounceButton() {
        Observable<Object> clickObservable = RxView.clicks(button);

        Observer<Object> clickObserver = new DisposableObserver<Object>() {
            @Override
            public void onNext(Object o) {
                logOnNext("Click");
            }

            @Override
            public void onError(Throwable e) {
                logOnError(e);
            }

            @Override
            public void onComplete() {
                logOnCompleted();
            }
        };

        clickObservable
                .debounce(1000, TimeUnit.MILLISECONDS)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(clickObserver);
    }

    private void reactiveService() {
        swapiService.getFilms()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new DisposableObserver<StarWarsFilms>() {
                    @Override
                    public void onNext(StarWarsFilms starWarsFilms) {
                        logOnNext("Found " + starWarsFilms.getResults().size() + " films");
                    }

                    @Override
                    public void onError(Throwable e) {
                        logOnError(e);
                    }

                    @Override
                    public void onComplete() {
                        logOnCompleted();
                    }
                });
    }

    private void reactiveServiceOneByOne() {
        swapiService.getFilms()
                .flatMap(new Function<StarWarsFilms, Observable<StarWarsFilm>>() {
                    @Override
                    public Observable<StarWarsFilm> apply(StarWarsFilms starWarsFilms) throws Exception {
                        return Observable.fromIterable(starWarsFilms.getResults());
                    }
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new DisposableObserver<StarWarsFilm>() {
                    @Override
                    public void onNext(StarWarsFilm starWarsFilm) {
                        logOnNext(starWarsFilm.getTitle());
                    }

                    @Override
                    public void onError(Throwable e) {
                        logOnError(e);
                    }

                    @Override
                    public void onComplete() {
                        logOnCompleted();
                    }
                });
    }

    private void logOnCompleted() {
        log("onCompleted");
    }

    private void logOnNext(String message) {
        log("onNext: " + message);
    }

    private void logOnError(Throwable error) {
        log("onError: " + error.getMessage());
    }

    private void log(String message) {
        log += message + "\n";
        logView.setText(log);
    }

}

File: app/src/main/java/be/appfoundry/aipdemo/activity/rxjava/ReactiveActivity.kt
package be.appfoundry.aipdemo.activity.rxjava

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import be.appfoundry.aipdemo.R
import be.appfoundry.aipdemo.extension.app
import be.appfoundry.aipdemo.service.SwapiService
import com.jakewharton.rxbinding2.view.RxView
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableObserver
import io.reactivex.rxkotlin.addTo
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_reactive.*
import java.util.concurrent.TimeUnit
import javax.inject.Inject

class ReactiveActivity : AppCompatActivity() {

    @Inject
    lateinit var swapiService: SwapiService

    private val disposables: CompositeDisposable = CompositeDisposable()

    private var log = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        app.appComponent.inject(this)

        setContentView(R.layout.activity_reactive)

        simpleRxJava()
        //simpleRxJavaWithOperators()
        //combinedObservables()
        //reactiveButton()
        //reactiveDebounceButton()
        //reactiveService()
        //reactiveServiceOneByOne()
    }

    override fun onDestroy() {
        super.onDestroy()

        disposables.clear()
    }

    private fun simpleRxJava() {
        val simpleStringObservable = Observable.just("Eric", "Koen", "Bart", "Lies", "Nikki")

        val simpleStringObserver = object : DisposableObserver<String>() {
            override fun onNext(s: String) {
                logOnNext(s)
            }

            override fun onError(e: Throwable) {
                logOnError(e)
            }

            override fun onComplete() {
                logOnCompleted()
            }
        }

        simpleStringObservable.subscribe(simpleStringObserver)

        disposables.add(simpleStringObserver)
    }

    private fun simpleRxJavaWithOperators() {
        Observable.just("Eric", "Koen", "Bart", "Lies", "Nikki")
                .filter { s -> s.contains("i") }
                .map { s -> s.toUpperCase() }
                .subscribe(
                        { s -> logOnNext(s) },
                        { e -> logOnError(e) },
                        { logOnCompleted() }
                )
                .addTo(disposables)
    }

    private fun combinedObservables() {
        val simpleMenObservable = Observable.just("Eric", "Koen", "Bart")
        val simpleWomenObservable = Observable.just("Lies", "Nikki")

        Observable.concat(simpleMenObservable, simpleWomenObservable)
                .subscribe(
                        { logOnNext(it) },
                        { logOnError(it) },
                        { logOnCompleted() }
                )
                .addTo(disposables)
    }

    private fun reactiveButton() {
        RxView.clicks(activityLogButton)
                .subscribe(
                        { logOnNext("Click") },
                        { logOnError(it) },
                        { logOnCompleted() }
                )
                .addTo(disposables)
    }

    private fun reactiveDebounceButton() {
        RxView.clicks(activityLogButton)
                .debounce(1000, TimeUnit.MILLISECONDS)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { logOnNext("Click") },
                        { logOnError(it) },
                        { logOnCompleted() }
                )
                .addTo(disposables)
    }

    private fun reactiveService() {
        swapiService.getFilms()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(
                        { logOnNext("Found " + it.results.size + " films") },
                        { logOnError(it) },
                        { logOnCompleted() }
                )
                .addTo(disposables)
    }

    private fun reactiveServiceOneByOne() {
        swapiService.getFilms()
                .flatMap { starWarsFilms -> Observable.fromIterable(starWarsFilms.results) }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(
                        { logOnNext(it.title) },
                        { logOnError(it) },
                        { logOnCompleted() }
                )
                .addTo(disposables)
    }

    private fun logOnCompleted() {
        log("onCompleted")
    }

    private fun logOnNext(message: String) {
        log("onNext: $message")
    }

    private fun logOnError(error: Throwable) {
        log("onError: ${error.message}")
    }

    private fun log(message: String) {
        log += message + "\n"
        activityLogText.text = log
    }
}